core/src/Kokkos_LogicalSpaces.hpp

```diff
@@ -0,0 +1,433 @@
+        s, "HostSpace", &s_root_record, detail);
+  }
+#else
+  static void print_records(std::ostream&, const SpaceType&,
+                            bool detail = false) {
+    (void)detail;
+    throw_runtime_exception(
+        "SharedAllocationRecord<HostSpace>::print_records only works "
+        "with "
+        "KOKKOS_ENABLE_DEBUG enabled");
+  }
+#endif
+};
+#ifdef KOKKOS_ENABLE_DEBUG
+/**\brief  Root record for tracked allocations from this LogicalSpace
+ * instance */
+template <class BaseSpace, class DefaultExecutionSpace, class Namer,
+          bool SharesAccessSemanticsWithBase>
+SharedAllocationRecord<void, void> SharedAllocationRecord<
+    Kokkos::Experimental::LogicalMemorySpace<
+        BaseSpace, DefaultExecutionSpace, Namer, SharesAccessSemanticsWithBase>,
+    void>::s_root_record;
+#endif
+
+}  // namespace Impl
+
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+
+namespace Impl {
+
+template <class Namer, class BaseSpace, class DefaultExecutionSpace,
+          bool SharesAccessSemanticsWithBase, class ExecutionSpace>
+struct DeepCopy<
+    Kokkos::Experimental::LogicalMemorySpace<
+        BaseSpace, DefaultExecutionSpace, Namer, SharesAccessSemanticsWithBase>,
+    Kokkos::Experimental::LogicalMemorySpace<
+        BaseSpace, DefaultExecutionSpace, Namer, SharesAccessSemanticsWithBase>,
+    ExecutionSpace> {
+  DeepCopy(void* dst, void* src, size_t n) {
+    DeepCopy<BaseSpace, BaseSpace, ExecutionSpace>(dst, src, n);
+  }
+  DeepCopy(const ExecutionSpace& exec, void* dst, void* src, size_t n) {
+    DeepCopy<BaseSpace, BaseSpace, ExecutionSpace>(exec, dst, src, n);
+  }
+};
+
+template <class Namer, class BaseSpace, class DefaultExecutionSpace,
+          bool SharesAccessSemanticsWithBase, class ExecutionSpace,
+          class SourceSpace>
+struct DeepCopy<
+    SourceSpace,
+    Kokkos::Experimental::LogicalMemorySpace<
+        BaseSpace, DefaultExecutionSpace, Namer, SharesAccessSemanticsWithBase>,
+    ExecutionSpace> {
+  DeepCopy(void* dst, void* src, size_t n) {
+    DeepCopy<SourceSpace, BaseSpace, ExecutionSpace>(dst, src, n);
+  }
+  DeepCopy(const ExecutionSpace& exec, void* dst, void* src, size_t n) {
+    DeepCopy<SourceSpace, BaseSpace, ExecutionSpace>(exec, dst, src, n);
+  }
+};
+
+template <class Namer, class BaseSpace, class DefaultExecutionSpace,
+          bool SharesAccessSemanticsWithBase, class ExecutionSpace,
+          class DestinationSpace>
+struct DeepCopy<
+    Kokkos::Experimental::LogicalMemorySpace<
+        BaseSpace, DefaultExecutionSpace, Namer, SharesAccessSemanticsWithBase>,
+    DestinationSpace, ExecutionSpace> {
+  DeepCopy(void* dst, void* src, size_t n) {
+    DeepCopy<BaseSpace, DestinationSpace, ExecutionSpace>(dst, src, n);
+  }
+  DeepCopy(const ExecutionSpace& exec, void* dst, void* src, size_t n) {
+    DeepCopy<BaseSpace, DestinationSpace, ExecutionSpace>(exec, dst, src, n);
+  }
+};
```

Comment:
Can you confirm that we want to provide a specialization even when `SharesAccessSemanticsWithBase` is `false`?

Reply:
I'll do you one better and confirm that we don't, and fix that

Reply:
We need them.

Reply:
Care to elaborate? I'd like to add a test that proves we need them, if we really do